design/one-pager-composition-environment-beta.md

```diff
@@ -0,0 +1,381 @@
+- use the `in-memory environment` to patch between composed resources without
+  setting up fields in the composite resource's status.
+- use the `in-memory environment` to temporarily hold data from different
+  sources (the composite resource and/or different composed resources) to
+  combine them in a subsequent patch.
+- select `EnvironmentConfig`s by label based on info from the Composite
+  resource.
+
+### Issues
+
+Early adopters' feedback clearly showed the following issues of the current
+implementation:
+
+- naming is confusing:
+    - expecting `ToEnvironment` patches to persist the state to some
+      `EnvironmentConfig`, showing confusion between `in-memory environment`
+      and `EnvironmentConfig`s.
+    - expecting the `environment` to be shared across all Composite resources
+      using a `Composition`, hence shared by all Composite Resources using a
+      specific `Composition`.
+- debugging is complex because of the following issues:
+    - confusion about the order in which patches are applied between the
+      Composite Resource, the `in-memory environment`, and Composed Resources.
+    - the lack of visibility of intermediate results for the `in-memory
+      environment`
+- the `environment`related part of the `Composition` API is perceived as overly
+  complex and error-prone, see
+  [https://github.com/crossplane/crossplane/issues/4738](https://github.com/crossplane/crossplane/issues/4738).
+- having to create an `EnvironmentConfig` to be able to access some information
+  from an independent Managed Resource from a `Composition` is considered to be
+  cumbersome.
+
+## Promotion to beta
+
+Given all the above, if we imagined splitting the functionality into its
+components and promoting them independently, we, the maintainers, would feel
+comfortable promoting to beta the following parts of the `Composition
+Environment` :
+
+- the `EnvironmentConfig` resource itself
+- the `in-memory environment` as an additional source and target for patches
+  defined at:
+    - `spec.resources[*].patches`
+    - `spec.environment.patches`, although these showed some discoverability
+      issues and could benefit some more thinking, also keeping into
+      consideration these are ignored for function-based Compositions.
+
+While we wouldn't feel so comfortable promoting the EnvironmentConfig selection
+part of the API at `spec.environment.environmentConfigs` in its current shape.
+
+Before promoting the entire feature to beta, it's essential to address the known
+issues and make the remaining parts of the Composition Environment more
+straightforward.
+
+### Independent Managed Resource referencing
+
+Currently, a Composition author who wanted to address some information from an
+existing Managed Resource would have the following options (simplified a bit
+the diagram not taking into account what the author knows about the existing
+MR):
+
+![beta-composition-environment-2][beta-composition-environment-2]
```

Comment:
Yes, exactly. It could actually be any known number of resources, the important bit is actually whether you know the resources beforehand or the actual resource to read is controlled by the end user of the composite resource. 

If you know them, today you can create 1 or more EnvironmentConfigs with the needed information and select the right one according to the user's input. 

If you don't know them, you can only compose an ObserveOnly resource according to the user input and therefore create one for each composite resource. 

I'll rephrase it to "Do you need to reference a finite set of Resources you know beforehand?", makes sense?

Reply:
thanks for the clarification. 

I like "Do you need to reference a finite set of Resources you know beforehand?". 

Some other suggestion:
"Do you need to reference a finite set of already known Resources?"

But your suggestion is good as well.